### ISSUE TITLE
constructor types not recognized if meta class used

### ISSUE THREAD

**Issue by matt-the-midnight-hacker** [microsoft/pyright#5586]
**Describe the bug**
I have a class with a constructor where a argument of certain type is expected e.g. `bool`. Pyright usually detects if I instantiate the class with a constructor argument of a wrong type. e.g. call it with a `string` instead of a `bool` type. If my class uses a metaclass which overwrites `__new__` and `__call__` then pyright can no longer detect the type of my class and suspects all arguments are from type `Any`.

**Code or Screenshots**

I hope this boiled down reproduction sample does dos not obscure the  issue I see in my actual code. I implemented here `__call__` and `__new__` minimal to trigger the behavior, My actual code does a little more, I believe type detection worked here once but not sure at which pyright version it stopped.

```python
class InitWithDefaults(type):
    def __call__(cls, *args: Any, **kwargs: Any):
        pass


class BaseMeta(InitWithDefaults):
    def __new__(mcs: Any, clsname: str):
        return type.__new__(mcs, clsname)


class Check(metaclass=BaseMeta):

    def __init__(self, success: bool) -> None:
        super().__init__()
        self.success: bool = success


c = Check(success="hallo")  # Here pyright does not complain about the wrong type

```

**VS Code extension or command-line**
I run pyright 1.1.318 in vscode.

**Comment by erictraut** [comment]
This was an intentional change. I recently rewrote the code in pyright that validates constructor calls. It now more closely reflects the runtime behavior. For details about how this works, refer to [this documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=constructor-calls).

If a metaclass overrides `type.__call__`, a type checker cannot make assumptions about the behavior of the metaclass. The normal behavior of `type.__call__` is to invoke the `__new__` and `__init__` method of the class, but metaclasses that override `__call__` may do something different. For example, sometimes `__call__` overrides are used to prevent explicit instantiation of a class, in which case they raise an exception.

Why is your metaclass overriding `type.__call__`? Does it call through to `type.__call__`? If so, does it return the value returned by `type.__call__`?

**Comment by matt-the-midnight-hacker** [comment]
Thank you for the fast answer and the information. The purpose in the code I work on is to "hook" the `__init__` function of each subclass to set default values for properties of specific types.

> Does it call through to `type.__call__`?

yes it does.

> does it return the value returned by `type.__call__`?

yes, it does after modifying the returned object.

If this is an intentional change, I guess I have to refactor this code then to have better type support.

**Comment by erictraut** [comment]
The reason I asked about your use case is that I'm considering changing the behavior in pyright for `type.__call__` overrides.

There's no way in the Python type system to indicate "I'm overriding `type.__call__`, but I'm preserving the semantics and behaviors of the original `type.__call__`. Pyright previously assumed that all overrides to `type.__call__` preserved the semantics unless they were annotated to return `NoReturn`. When I rewrote the constructor evaluation logic, I changed it to conservatively assume that no override to `type.__call__` preserved the behavior of `type.__call__`. That might be too conservative, so I've been considering a heuristic that represents a middle ground. If the override returns the same type as `type.__call__` would have, it may be safe(ish) to assume that the override preserves the same behaviors of `type.__call__`. That would restore the older behavior in your case. What do you think?

It sounds like your metaclass is doing additional hackery (dynamically injecting new instance variables?) that a static code analyzer will not have insight into. I don't have a good solution for that.

**Comment by matt-the-midnight-hacker** [comment]
The change you considering would restore the older behavior in my case. I'd like that. 

> It sounds like your metaclass is doing additional hackery

Yes somehow it does. The  `__new__` implementation creates dynamically marshmallow schemas to dump json data. But it does not dynamically inject new instance variables. The `__call__` implementation sets default values on None values for specific types. All this could be surely also implemented different on my end but that's how things currently work.

**Comment by erictraut** [comment]
I've updated pyright's heuristics as I suggested above. There's no perfect solution here, but I'm hopeful that the new heuristics will produce better results, on average, that the previous heuristics. This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.320, which I just published. It will also be included in a future version of pylance.
